src/effects/VST3/VST3Effect.cpp

```diff
@@ -0,0 +1,1138 @@
+      return true;
+   }
+   catch(std::exception& e)
+   {
+      //make_unique<> isn't likely to fail here since this effect instance
+      //somehow succeeded
+      wxLogError("Failed to add realtime processor: %s", e.what());
+   }
+   return false;
+}
+
+bool VST3Effect::RealtimeFinalize()
+{
+   for(auto& processor : mRealtimeGroupProcessors)
+      processor->ProcessFinalize();
+
+   mRealtimeGroupProcessors.clear();
+   mPendingChanges.reset();
+   
+   return true;
+}
+
+bool VST3Effect::RealtimeSuspend()
+{
+   for(auto& effect : mRealtimeGroupProcessors)
+      effect->mAudioProcessor->setProcessing(false);
+   return true;
+}
+
+bool VST3Effect::RealtimeResume()
+{
+   for(auto& effect : mRealtimeGroupProcessors)
+      effect->mAudioProcessor->setProcessing(true);
+   return true;
+}
+
+bool VST3Effect::RealtimeProcessStart()
+{
+   wxASSERT(mPendingChanges == nullptr);
+
+   if(mComponentHandler != nullptr)
+      //Same parameter changes are used among all of the relatime processors
+      mPendingChanges = mComponentHandler->getPendingChanges();
+   return true;
+}
+
+size_t VST3Effect::RealtimeProcess(int group, float** inBuf, float** outBuf, size_t numSamples)
+{
+   auto& effect = mRealtimeGroupProcessors[group];
+   return VST3ProcessBlock(effect->mEffectComponent.get(), effect->mSetup, inBuf, outBuf, numSamples, mPendingChanges.get());
+}
+
+bool VST3Effect::RealtimeProcessEnd()
+{
+   mPendingChanges.reset();
+   return true;
+}
+
+int VST3Effect::ShowClientInterface(wxWindow& parent, wxDialog& dialog, bool forceModal)
+{
+   if(!IsGraphicalUI())
+   {
+      //Restrict resize of the "plain" dialog
+      dialog.SetMaxSize(dialog.GetSize());
+      dialog.SetMinSize(dialog.GetSize());
+   }
+   if(forceModal)
+      return dialog.ShowModal();
+
+   dialog.Show();
+   return 0;
+}
+
+bool VST3Effect::SetHost(EffectHostInterface* host)
+{
+   mEffectHost = host;
+   return true;
+}
+
+bool VST3Effect::IsGraphicalUI()
+{
+   return mPlugView != nullptr;
+}
+
+bool VST3Effect::PopulateUI(ShuttleGui& S)
+{
+   using namespace Steinberg;
+
+   mParent = S.GetParent();
+
+   if(mComponentHandler != nullptr)
+   {
+      SyncParameters();
+
+      auto parent = S.GetParent();
+      if(GetType() == EffectTypeGenerate)
+      {
+         auto vSizer = std::make_unique<wxBoxSizer>(wxVERTICAL);
+         auto controllsRoot = std::make_unique<wxWindow>(parent, wxID_ANY);
+         if(!LoadVSTUI(controllsRoot.get()))
+            VST3Utils::BuildPlainUI(controllsRoot.get(), mEditController, mComponentHandler);
+         vSizer->Add(controllsRoot.release());
+
+         auto duration = std::make_unique<NumericTextCtrl>(
+               parent, wxID_ANY,
+               NumericConverter::TIME,
+               mEffectHost->GetDurationFormat(),
+               mEffectHost->GetDuration(),
+               mSetup.sampleRate,
+               NumericTextCtrl::Options{}
+                  .AutoPos(true)
+            );
```

Comment:
Complete the analogy with `VSTEffect::BuildPlain`.  Do not omit the `SetName` call.  David Bailes will tell you it is important so that the screen reader pronounces the name.